### PR TITLE
[#IC-193] Add `io-subscription-migrations` function app

### PR DESCRIPTION
**Pull request by balanza**
### List of changes
 

* define subnet `fnsubscriptionmigrations`
* define function app `io-subscription-migrations`
* define `staging` slot

### Motivation and context

This is a new application under `selfcare` domain (with which it shares app plan and resource group)

### Type of changes

- [x] Add new resources
- [ ] Update configuration to existing resources
- [ ] Remove existing resources

### Env to apply

- [ ] DEV
- [ ] UAT
- [x] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [x] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [x] No

### Other information

---

### CODE REVIEW

File: src/core/function_subscription_migrations.tf [hunk 1]
```diff
@@ -0,0 +1,92 @@
+module "function_subscriptionmigrations_snet" {
+  source               = "git::https://github.com/pagopa/azurerm.git//subnet?ref=v2.1.0"
+  name                 = "fnsubscriptionmigrations"
```
Comment (uolter) [inline src/core/function_subscription_migrations.tf:3]:
```suggestion
  name                 = format("%s-fn-sub-migrations-snet", local.project)
```

File: src/core/function_subscription_migrations.tf [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    FETCH_KEEPALIVE_TIMEOUT             = "60000"
+  }
+}
+
+module "function_subscriptionmigrations" {
+  source = "git::https://github.com/pagopa/azurerm.git//function_app?ref=v2.1.0"
+
+  application_insights_instrumentation_key = data.azurerm_application_insights.application_insights.instrumentation_key
+  location                                 = var.location
+  name                                     = "subscriptionmigrations"
```
Comment (uolter) [inline src/core/function_subscription_migrations.tf:43]:
```suggestion
  name                                     = format("%s-subscription-migrations-fn", local.project)
```

**Review by uolter — CHANGES_REQUESTED** [review]

File: src/core/function_subscription_migrations.tf [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  resource_group_name                      = azurerm_resource_group.selfcare_be_rg.name
+  subnet_id                                = module.function_subscriptionmigrations_snet.id
+  tags                                     = var.tags
+  allowed_ips                              = local.app_insights_ips_west_europe
+  allowed_subnets = [
+    data.azurerm_subnet.azdoa_snet[0].id,
+  ]
+
+  app_service_plan_name = azurerm_app_service_plan.selfcare_be_common.name
```
Comment (pasqualedevita) [inline src/core/function_subscription_migrations.tf:52]: you need to use app_service_plan_id, plan name is for new app service plan

File: src/core/function_subscription_migrations.tf [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  ]
+
+  delegation = {
+    name = "default"
+    service_delegation = {
+      name    = "Microsoft.Web/serverFarms"
+      actions = ["Microsoft.Network/virtualNetworks/subnets/action"]
+    }
+  }
+}
```
Comment (pasqualedevita) [inline src/core/function_subscription_migrations.tf:19]: when you share app service plans u need to reuse associated subnet, so u need to remove this subnet

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: src/core/function_subscription_migrations.tf (hunk 4 above)
Comment (balanza) [reply src/core/function_subscription_migrations.tf]: Tried to fix with this: https://github.com/pagopa/io-infra/pull/77/commits/92520c88d225ef06b2540812477ad81d97c28b09

**Comment by balanza** [comment]
@pasqualedevita @uolter in the end, I figured out almost everything, except:
> Error: Invalid value for module argument
│ 
│   on function_subscription_migrations.tf line 43, in module "function_subscriptionmigrations":
│   43:   internal_storage = {
│   44:     "blobs_retention_days" : 1,
│   45:     "containers" : [],
│   46:     "enable" : true,
│   47:     "queues" : []
│   48:   }
│ 
│ The given value is not suitable for child module variable
│ "internal_storage" defined at
│ .terraform/modules/function_subscriptionmigrations/function_app/variables.tf:118,1-28:
│ attributes "private_dns_zone_blob_ids", "private_dns_zone_queue_ids",
│ "private_dns_zone_table_ids", and "private_endpoint_subnet_id" are
│ required.

I don't actually know what to put in there. Can I infer them from somewhere else?

**Comment by uolter** [comment]
> @pasqualedevita @uolter in the end, I figured out almost everything, except:
> 
> > Error: Invalid value for module argument
> > │
> > │   on function_subscription_migrations.tf line 43, in module "function_subscriptionmigrations":
> > │   43:   internal_storage = {
> > │   44:     "blobs_retention_days" : 1,
> > │   45:     "containers" : [],
> > │   46:     "enable" : true,
> > │   47:     "queues" : []
> > │   48:   }
> > │
> > │ The given value is not suitable for child module variable
> > │ "internal_storage" defined at
> > │ .terraform/modules/function_subscriptionmigrations/function_app/variables.tf:118,1-28:
> > │ attributes "private_dns_zone_blob_ids", "private_dns_zone_queue_ids",
> > │ "private_dns_zone_table_ids", and "private_endpoint_subnet_id" are
> > │ required.
> 
> I don't actually know what to put in there. Can I infer them from somewhere else?

@balanza I'll give it a go tomorrow. .... I think we need to create the private links before.
something like we did here:
https://github.com/pagopa/io-infrastructure-live-new/blob/master/prod/westeurope/common/private_dns_zones/privatelink-queue-core-windows-net/zone/terragrunt.hcl
Instead of the terragrunt code we can use the terraform code like the one inside that module.

File: src/core/function_subscription_migrations.tf [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  ]
+
+  app_service_plan_id = local.app_context.app_service_plan.id
+  health_check_path   = "api/v1/info"
+  internal_storage = {
+    "blobs_retention_days" : 1,
+    "containers" : [],
+    "enable" : true,
+    "queues" : []
+  }
```
Comment (pasqualedevita) [inline src/core/function_subscription_migrations.tf:48]:
```suggestion
  internal_storage = {
    "enable"                     = true,
    "private_endpoint_subnet_id" = data.azurerm_subnet.private_endpoints_subnet.id,
    "private_dns_zone_blob_ids"  = [data.azurerm_dns_zone.privatelink_blob_core_windows_net.id],
    "private_dns_zone_queue_ids" = [data.azurerm_dns_zone.privatelink_queue_core_windows_net.id],
    "private_dns_zone_table_ids" = [data.azurerm_dns_zone.privatelink_table_core_windows_net.id],
    "queues"                     = [],
    "containers"                 = [],
    "blobs_retention_days"       = 1,
  } 
```

Also u need to add data to query:
- private_endpoint_subnet_id
- private_dns_zone_blob_ids
- private_dns_zone_queue_ids
- private_dns_zone_table_ids

```suggestion
# data.tf get private_endpoint_subnet_id 
data "azurerm_subnet" "private_endpoints_subnet" {
  name                 = "pendpoints"
  virtual_network_name = format("%s-vnet-common", local.project)
  resource_group_name  = format("%s-rg-common", local.project)
}

# private_dns_zones.tf
data "azurerm_dns_zone" "privatelink_blob_core_windows_net" {
  name                = "privatelink.blob.core.windows.net"
  resource_group_name  = format("%s-rg-common", local.project)
}

data "azurerm_dns_zone" "privatelink_queue_core_windows.net" {
  name                = "privatelink.queue.core.windows.net"
  resource_group_name  = format("%s-rg-common", local.project)
}

data "azurerm_dns_zone" "privatelink_file_core_windows.net" {
  name                = "privatelink.file.core.windows.net"
  resource_group_name  = format("%s-rg-common", local.project)
}

data "azurerm_dns_zone" "privatelink_table_core_windows.net" {
  name                = "privatelink.table.core.windows.net"
  resource_group_name  = format("%s-rg-common", local.project)
}

```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: src/core/function_subscription_migrations.tf (hunk 5 above)
Comment (balanza) [reply src/core/function_subscription_migrations.tf]: done in https://github.com/pagopa/io-infra/pull/77/commits/a40ccb10ddc544516c3289c02ecad4e942d399b7

Comment (pasqualedevita) [reply src/core/function_subscription_migrations.tf]: @balanza my bad is azurerm_private_dns_zone not azurerm_dns_zone.

also I see a typo `privatelink_table_core_windows.net` -> `privatelink_table_core_windows_net`

File: src/core/function_subscription_migrations.tf [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  ]
+
+  app_service_plan_id = local.app_context.app_service_plan.id
+  health_check_path   = "api/v1/info"
+  internal_storage = {
+    "enable"                     = true,
+    "private_endpoint_subnet_id" = data.azurerm_subnet.private_endpoints_subnet.id,
+    "private_dns_zone_blob_ids"  = [data.azurerm_dns_zone.privatelink_blob_core_windows_net.id],
+    "private_dns_zone_queue_ids" = [data.azurerm_dns_zone.privatelink_queue_core_windows_net.id],
+    "private_dns_zone_table_ids" = [data.azurerm_dns_zone.privatelink_table_core_windows_net.id],
```
Comment (pasqualedevita) [inline src/core/function_subscription_migrations.tf:48]:
```suggestion
    "private_dns_zone_blob_ids"  = [data. azurerm_private_dns_zone.privatelink_blob_core_windows_net.id],
    "private_dns_zone_queue_ids" = [data. azurerm_private_dns_zone.privatelink_queue_core_windows_net.id],
    "private_dns_zone_table_ids" = [data. azurerm_private_dns_zone.privatelink_table_core_windows_net.id],
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: src/core/function_subscription_migrations.tf [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    FETCH_KEEPALIVE_TIMEOUT             = "60000"
+  }
+
+  // As we run this application under SelfCare IO logic subdomain, 
+  //  we share some resources
+  app_context = {
+    resource_group   = azurerm_resource_group.selfcare_be_rg
+    app_service_plan = azurerm_app_service_plan.selfcare_be_common
+    snet             = module.selfcare_be_common_snet
+  }
```
Comment (pasqualedevita) [inline src/core/function_subscription_migrations.tf:23]: locals are global so it's better to define inner object, for example:

```
locals {

 function_subscriptionmigrations = {
  app_context = {
  }
  app_settings_commons = {
  }
 } 


}
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: src/core/function_subscription_migrations.tf (hunk 7 above)
Comment (balanza) [reply src/core/function_subscription_migrations.tf]: isn't there a file-scoped way to declare variables?

Comment (pasqualedevita) [reply src/core/function_subscription_migrations.tf]: no :( terraform reads all files in a folder

**Review by uolter — APPROVED** [review]

**Pull request merged**